libs/ledger/src/chain/main_chain.cpp

```diff
@@ -1442,23 +1621,70 @@ bool MainChain::ReindexTips()
       continue;
     }
     // this hash has no next blocks
-    auto const &   block{*block_entry.second};
-    const uint64_t total_weight{block.total_weight};
-    const uint64_t weight{block.weight};
-    const uint64_t block_number{block.block_number};
-    new_tips[hash] = Tip{total_weight, weight};
-    // check if this tip is the current heaviest
-    if (total_weight > max_total_weight ||
-        (total_weight == max_total_weight && block_number > max_block_number) ||
-        (total_weight == max_total_weight && block_number == max_block_number &&
-         weight > max_weight) ||
-        (total_weight == max_total_weight && block_number == max_block_number &&
-         weight == max_weight && hash > max_hash))
+    auto &      block{*block_entry.second};
+    BlockWeight total_weight{block.total_weight};
+    BlockWeight weight{block.weight};
+    BlockNumber block_number{block.block_number};
+
+    // check if block is stutter block and if so loop until a non-stutter block is found and replace
+    // as tip
+    bool replaced{true};
+    if (enable_stutter_removal_)
     {
-      max_total_weight = total_weight;
-      max_weight       = weight;
-      max_hash         = hash;
-      max_block_number = block_number;
+      assert(stutter_blocks_.find(block_number) != stutter_blocks_.end() &&
+             stutter_blocks_.at(block_number).find(weight) !=
+                 stutter_blocks_.at(block_number).end());
+      if (stutter_blocks_[block_number][weight])
+      {
+        replaced            = false;
+        auto previous_block = GetBlock(block.previous_hash);
+
+        while (!replaced)
+        {
+          BlockNumber previous_block_number = previous_block->block_number;
+          assert(stutter_blocks_.find(previous_block_number) != stutter_blocks_.end() &&
+                 stutter_blocks_.at(previous_block_number).find(previous_block->weight) !=
+                     stutter_blocks_.at(previous_block_number).end());
+          if (previous_block->IsGenesis() ||
+              !stutter_blocks_[previous_block_number][previous_block->weight])
+          {
+            block        = *previous_block;
+            hash         = block.hash;
+            total_weight = block.total_weight;
+            weight       = block.weight;
+            block_number = block.block_number;
+            replaced     = true;
+          }
+          else
+          {
+            previous_block = GetBlock(previous_block->previous_hash);
+            if (!previous_block)
+            {
+              FETCH_LOG_WARN(LOGGING_NAME, "Main chain failed to replace stutter tip previous");
+              break;
+            }
+          }
+        }
+      }
+    }
+
+    if (replaced)
+    {
+      new_tips[hash] = Tip{total_weight, weight, block_number};
```

Comment:
Just below this, there is:

```
      // check if this tip is the current heaviest
      if (total_weight > max_total_weight ||
          (total_weight == max_total_weight && block_number > max_block_number) ||
          (total_weight == max_total_weight && block_number == max_block_number &&
           weight > max_weight) ||
          (total_weight == max_total_weight && block_number == max_block_number &&
           weight == max_weight && hash > max_hash))
```

this should probably be refactored - I think the whole function could become something of the form:

```
// Clear old tips and repopulate
tips_.clear();
std::set<Block> block_tips;

for (auto const &block_entry : block_chain_)
{
  // Loose blocks can't be tips
  if (block_entry.second->is_loose)
  {
    continue;
  }

  // Blocks with a forward reference can't be tips (exception for stutter blocks)
  if(HasForwardRef(block_entry))
  {
    continue;
  }

  // If this is a stutter block, it walk down until we find a non-stutter to use as the tip (otherwise insert block_entry)
  block_tips.insert(GetFirstNonStutter(block_entry));
}

// just for safety
if(tips_.empty())
{
  FETCH_LOG_WARN(LOGGING_NAME, "tips empty somehow!");
  AddTip(genesis);
}

// Sanity check - is heaviest_ in tips_?

// Loop here populating tips_ with block_tips
//...

// Set heaviest_
}
```

Reply:
Doing it this way you would have to loop through the set of tips again to find the heaviest, no? Whereas as it is now it is determined when a tip is added to new_tips so it can just be set at the end.